pinot-broker/src/main/java/org/apache/pinot/broker/querylog/QueryLogger.java

```diff
@@ -0,0 +1,294 @@
+    _enableIpLogging = enableIpLogging;
+    _logger = logger;
+    _droppedLogRateLimiter = droppedLogRateLimiter;
+  }
+
+  public void log(QueryLogParams params) {
+    _logger.debug("Broker Response: {}", params._response);
+
+    if (!(_logRateLimiter.tryAcquire() || shouldForceLog(params))) {
+      _numDroppedLogs.incrementAndGet();
+      return;
+    }
+
+    final StringBuilder queryLogBuilder = new StringBuilder();
+    for (QueryLogEntry value : QueryLogEntry.values()) {
+      value.format(queryLogBuilder, this, params);
+      queryLogBuilder.append(',');
+    }
+
+    // always log the query last - don't add this to the QueryLogEntry enum
+    queryLogBuilder.append("query=")
+        .append(StringUtils.substring(params._query, 0, _maxQueryLengthToLog));
+    _logger.info(queryLogBuilder.toString());
+
+    if (_droppedLogRateLimiter.tryAcquire()) {
+      // use getAndSet to 0 so that there will be no race condition between
+      // loggers that increment this counter and this thread
+      long numDroppedLogsSinceLastLog = _numDroppedLogs.getAndSet(0);
+      if (numDroppedLogsSinceLastLog > 0) {
+        _logger.warn("{} logs were dropped. (log max rate per second: {})",
+            numDroppedLogsSinceLastLog,
+            _droppedLogRateLimiter.getRate());
+      }
+    }
+  }
+
+  public int getMaxQueryLengthToLog() {
+    return _maxQueryLengthToLog;
+  }
+
+  public double getLogRateLimit() {
+    return _logRateLimiter.getRate();
+  }
+
+  private boolean shouldForceLog(QueryLogParams params) {
+    return params._response.isNumGroupsLimitReached()
+        || params._response.getExceptionsSize() > 0
+        || params._timeUsedMs > TimeUnit.SECONDS.toMillis(1);
+  }
+
+  public static class QueryLogParams {
+    final long _requestId;
+    final String _query;
+    final RequestContext _requestContext;
+    final String _table;
+    final int _numUnavailableSegments;
+    final BaseBrokerRequestHandler.ServerStats _serverStats;
+    final BrokerResponse _response;
+    final long _timeUsedMs;
+    @Nullable final RequesterIdentity _requester;
+
+    public QueryLogParams(
+        long requestId,
+        String query,
+        RequestContext requestContext,
+        String table,
+        int numUnavailableSegments,
+        BaseBrokerRequestHandler.ServerStats serverStats,
+        BrokerResponse response,
+        long timeUsedMs,
+        @Nullable RequesterIdentity requester
+    ) {
+      _requestId = requestId;
+      _query = query;
+      _table = table;
+      _timeUsedMs = timeUsedMs;
+      _requestContext = requestContext;
+      _requester = requester;
+      _response = response;
+      _serverStats = serverStats;
+      _numUnavailableSegments = numUnavailableSegments;
+    }
+  }
+
+  /**
+   * NOTE: please maintain the order of this query log entry enum. If you want to add a new
+   * entry, add it to the end of the existing list.
+   */
+  private enum QueryLogEntry {
+    REQUEST_ID("requestId") {
```

Comment:
I think my previous question is addressed by this. Seems like the order is same as today.